### PR TITLE
Binfhe extended

### PR DESCRIPTION
**Pull request by pascoec**
- Added extended functionality to EvalBinGate() and Bootstrapping()
- Added unit tests for extended functionality
- Added move constructors for many classes in binfhe
- Bug fix for binfhe pk encryption
- Updates to binfhe-paramsets benchmark
- Update to operator<<(std::ostream& s, BINFHE_OUTPUT f)
- Changed default BINFHE_OUTPUT for secret key encryption to SMALL_DIM
- Limit use of FRESH and BOOTSTRAPPED BINFHE_OUTPUT

### CODE REVIEW

**Comment by yspolyakov** [comment]
@pascoec I suggest running the main binfhe benchmark to see how the runtime changes with the PR (before and after).

**Comment by pascoec** [comment]
icelake, Clang17, WITH_OPENMP=OFF, WITH_NATIVEOPT=ON, NATIVE_SIZE=32

dev:
------------------------------------------------------------------------------------------------
Benchmark                                                      Time             CPU   Iterations
------------------------------------------------------------------------------------------------
FHEW_BINGATE2/TOY_2_GINX_OR/min_time:5.000                  2.12 ms         2.12 ms         3288
FHEW_BINGATE2/MEDIUM_2_GINX_OR/min_time:5.000               27.6 ms         27.6 ms          255
FHEW_BINGATE2/STD128_2_AP_OR/min_time:5.000                 51.3 ms         51.3 ms          136
FHEW_BINGATE2/STD128_2_GINX_OR/min_time:5.000               32.9 ms         32.9 ms          213
FHEW_BINGATE3/STD128_3_GINX_OR/min_time:5.000               53.0 ms         53.0 ms          133
FHEW_BINGATE4/STD128_4_GINX_OR/min_time:5.000               53.0 ms         53.0 ms          133
FHEW_BINGATE2/STD128Q_2_GINX_OR/min_time:5.000              47.1 ms         47.1 ms          148
FHEW_BINGATE2/STD256Q_2_GINX_OR/min_time:5.000               215 ms          215 ms           32
FHEW_BINGATE3/STD256Q_3_GINX_OR/min_time:5.000               314 ms          314 ms           22
FHEW_BINGATE4/STD256Q_4_GINX_OR/min_time:5.000               366 ms          366 ms           19
FHEW_BINGATE2/STD128_2_LMKCDEY_OR/min_time:5.000            36.0 ms         36.0 ms          195
FHEW_BINGATE3/STD128_3_LMKCDEY_OR/min_time:5.000            43.8 ms         43.8 ms          160
FHEW_BINGATE4/STD128_4_LMKCDEY_OR/min_time:5.000            62.8 ms         62.8 ms          112
FHEW_BINGATE2/STD128Q_2_LMKCDEY_OR/min_time:5.000           38.6 ms         38.6 ms          179
FHEW_BINGATE3/STD128Q_3_LMKCDEY_OR/min_time:5.000           67.0 ms         66.9 ms          104
FHEW_BINGATE2/LPF_STD128_2_GINX_OR/min_time:5.000           49.5 ms         49.5 ms          143
FHEW_BINGATE2/LPF_STD128Q_2_GINX_OR/min_time:5.000          57.1 ms         57.1 ms          122
FHEW_BINGATE2/LPF_STD128_2_LMKCDEY_OR/min_time:5.000        43.6 ms         43.6 ms          160
FHEW_BINGATE2/LPF_STD128Q_2_LMKCDEY_OR/min_time:5.000       62.9 ms         62.9 ms          111

binfhe_extended:
------------------------------------------------------------------------------------------------
Benchmark                                                      Time             CPU   Iterations
------------------------------------------------------------------------------------------------
FHEW_BINGATE2/TOY_2_GINX_OR/min_time:5.000                  2.12 ms         2.12 ms         3297
FHEW_BINGATE2/MEDIUM_2_GINX_OR/min_time:5.000               27.5 ms         27.5 ms          258
FHEW_BINGATE2/STD128_2_AP_OR/min_time:5.000                 51.5 ms         51.5 ms          134
FHEW_BINGATE2/STD128_2_GINX_OR/min_time:5.000               32.7 ms         32.7 ms          213
FHEW_BINGATE3/STD128_3_GINX_OR/min_time:5.000               52.8 ms         52.8 ms          132
FHEW_BINGATE4/STD128_4_GINX_OR/min_time:5.000               52.7 ms         52.7 ms          132
FHEW_BINGATE2/STD128Q_2_GINX_OR/min_time:5.000              47.1 ms         47.1 ms          147
FHEW_BINGATE2/STD256Q_2_GINX_OR/min_time:5.000               218 ms          218 ms           32
FHEW_BINGATE3/STD256Q_3_GINX_OR/min_time:5.000               315 ms          315 ms           22
FHEW_BINGATE4/STD256Q_4_GINX_OR/min_time:5.000               368 ms          368 ms           19
FHEW_BINGATE2/STD128_2_LMKCDEY_OR/min_time:5.000            36.0 ms         36.0 ms          194
FHEW_BINGATE3/STD128_3_LMKCDEY_OR/min_time:5.000            44.0 ms         44.0 ms          159
FHEW_BINGATE4/STD128_4_LMKCDEY_OR/min_time:5.000            63.0 ms         63.0 ms          112
FHEW_BINGATE2/STD128Q_2_LMKCDEY_OR/min_time:5.000           38.6 ms         38.6 ms          181
FHEW_BINGATE3/STD128Q_3_LMKCDEY_OR/min_time:5.000           67.2 ms         67.2 ms          103
FHEW_BINGATE2/LPF_STD128_2_GINX_OR/min_time:5.000           49.2 ms         49.2 ms          142
FHEW_BINGATE2/LPF_STD128Q_2_GINX_OR/min_time:5.000          57.1 ms         57.1 ms          121
FHEW_BINGATE2/LPF_STD128_2_LMKCDEY_OR/min_time:5.000        43.8 ms         43.8 ms          159
FHEW_BINGATE2/LPF_STD128Q_2_LMKCDEY_OR/min_time:5.000       63.1 ms         63.1 ms          111

File: src/binfhe/unittest/UnitTestFHEWExtended.cpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    auto pk = cc.GetPublicKey();
+    auto Q  = cc.GetParams()->GetLWEParams()->GetQ();
+
+    auto ct_small = cc.Encrypt(pk, 1, SMALL_DIM, 8);
+    EXPECT_NE(Q, ct_small->GetModulus());
+
+    auto ct_large = cc.Encrypt(pk, 1, LARGE_DIM, 8);
+    EXPECT_EQ(Q, ct_large->GetModulus());
+
+    std::vector<LWECiphertext> v{ct_small, ct_large, cc.Encrypt(pk, 0, FRESH, 8), cc.Encrypt(pk, 1, FRESH, 8)};
```
[the remark this answers is not in the document]
Comment (pascoec) [reply src/binfhe/unittest/UnitTestFHEWExtended.cpp]: The default output for secret key encryption is BOOTSTRAPPED. Should I change it to LARGE_DIM?

[the remark this answers is not in the document]
Comment (yspolyakov) [reply src/binfhe/unittest/UnitTestFHEWExtended.cpp]: No, it should be SMALL_DIM

[the remark this answers is not in the document]
Comment (pascoec) [reply src/binfhe/unittest/UnitTestFHEWExtended.cpp]: changed

**Review by yspolyakov — APPROVED** [review]

**Pull request merged**